### ISSUE TITLE
FR: upload and download xdrip config from / to central server

### ISSUE THREAD

**Issue by gruoner** [NightscoutFoundation/xDrip#1661]
after outage of my mobile i needed to restore the whole xdrip config (data collector, uploads, alarms....) and that was really a lot of work. So i want to store (and restore if needed) the whole xdrip config (and maybe the whole database) on a central config server.

Whenever a preferences change occurs an upload should be triggered to send the actual config after x minutes without preference-changes - that's also easy to implement as all preference changes are sent throu static Pref class and their setters....

As far as i can see there seems to be two options i want to discuss:

1. a WebDAV client (based on sardine-android - https://github.com/thegrizzlylabs/sardine-android) to up- and download config (and even the whole database) via webdav
2. extend nightscout server to get a place where one can upload and download the config (database is already uploaded there). Last month i wrote an extension to nightscout server; thats quite easy - its more complicated to merge it into master branch

Both options doesn't seems much complicated but which one would be more "useful" for others?

**Comment by Navid200** [comment]
The best way to keep my blood glucose readings safe is Nightscout.   For example, if I lose my phone, all my glucose history, that is on the phone, will not be lost.

If I backup my database, I will not be able to impost it.  If I do, it will overwrite what I have.  So, unless I import it into a phone that I have never used, it's useless.  And if I import a database I saved 2 days ago, obviously, it will only recover all my readings up to 2 days ago.  Nightscout, however, will have all my readings that I have uploaded.

There are xDrip settings that are device-specific.  There are settings one has to set if one has a Samsung and unset if one doesn't.  So, it would be pointless to import such settings saved from one phone into another phone.

The only time I can think of when I may benefit from having a backup of my settings is if I do significant customizations to alarms, display coloring, etc. or for creating a backup of long not-easy to copy settings perhaps like Nightscout login and password etc.
But, those are not things I would change on a daily basis.  I would change and change and change until I have what I like.  Of course, then, it makes sense to make a backup.  And I can easily do that already using the existing Import/Export features.

I don't see a reason to change anything.
Can you explain a reason you need to backup your settings every time you make a single change?

**Comment by gruoner** [comment]
backing up database is just for users not using any cloud-upload-option - for all the others its useless; i agree
but i see your point with device-specific config; as i'm always using samsung i couldn't see any differences while copying a really old backup to the new phone (newer version of xdrip...).
The use case a had last weeks was quite simple: my phone broke up and i didn't have an actual backup (as this is a xdrip companion and not a daily working smartphone i've never done a backup; complete diab diary -the real important thing- is on nightscout server). But in the last months i adjusted several configs (esp. G6 transmitter number, sensor codes, alarm settings, graph setting, prediction setting and some more i can't remember...). As i'm running a special feature branch for a bit more sophisticated alarm setting (i will develop it in direction of dexcom's alarm plans) i also had to redefine all the alarming config.
-- it took me a week to recap all config and i'm not really sure if i got all of them.

Config will change over weeks and months and we are getting more and more config in xdrip (nearly every feature needs a switch to switch it off or to configure the feature). So i want an easy way to backup the config whenever some config changes and also restore it when i have to move to another mobile. i can install a slim config just for restore together with my xdrip installation (i'm running my own xdrip stream); but thats just a thought.....

As i will implement this feature (because i need it) the question is just about interests of other xdrip users and whether i will rise a PR at the end or not. I experienced hard bricks to get a feature into xdrip master and i want to know if someone else is also interested in such a feature.....

**Comment by Navid200** [comment]
I wish we could simplify the xDrip code instead of making it even more complicated than it already is.
You are right, there are too many settings.
If we keep approving feature requests, it is only going to get worse.

Your solution is to make it easier to backup the settings so that having too many settings is not a problem any longer.
I have to disagree.  I don't disagree that it will make backing it up easier.
But, the main problem with having so many settings is not backing it up.  The problem is that a user doesn't know how to use xDrip.

I spend most of my time on facebook walking users through setting up G5 or G6.  Everyday, there is a user showing a G6 with stopped sensor and they don't know what to do.
I am helping someone right now as I am typing this.

PLEASE stop adding settings to xDrip.  Please only approve new features that are absolutely necessary.  Please reject any feature request that has a work-around.

You can already backup your settings.  From the top right menu button, tap on Impost/Export features.  Then, tap on Load / Save Settings on sdcard.  Then, tap on Save all settings to sdcard.
Now, go to your file manager in Android and go to the Downloads folder and find the backup.  You can save it in a safe place on your server.  It takes less than 2 minutes to do all of that.

I am not suggesting that your proposal has no value.  All I am saying is that there are pros and cons with approving new features.  

Also, we already have so many open issues.
Someone asked for the missed reading alert to have an override silent mode option added to it.  The request was made 3.5 years ago.
https://github.com/NightscoutFoundation/xDrip/issues/193
This is a legitimate request.  I see the request on facebook too.  Someone wakes up in the morning having a hypo.  Later, realizes that there was no alarm because there were no readings but the missed reading alarm did not go off because of DND.  We need to address that request.
As long as we have open issues like that open, I don't see a reason to spend time on anything else.  
Can someone explain to me why any pull request we have open right now is more critical than issue 193?

**Comment by gruoner** [comment]
i see your point but have to totally disagree - after rising my PR for multi insulin i had have a discussion with jamorham whether we want to start a new release train with a dedicated feature set for "more sophisticated users". his vote was really clear - no new train.
This flood of feature switches is a direct consequence and we are getting lots of more features and switches - at least i'm running 4 feature branches (and have 2 new in mind for features i'm missing) and i added nearly a dozen new config prefs. I agree: its is getting worse.
**BUT** do u really want to ignore the demand of new features and do u really want to upset all the developers evolving xdrip? When #193 is such a big issue, such an important feature why isn't it still implemented? Community developers are implementing things **they** need; if it matches demands of other users i'm willing to address them whenever they doesn't collide with mine - maybe #193 can be a FR i can address with my ideas concerning alarm plans like dexcom has recently implemented.....

I'm in a very luxury position: i can run my own train, i can implement and include new features and branches on my needs and demands. BUT also a dedicated train just on my own has limits - every change in the master branch might result in huge efforts merging them to my feature branches. I run dedicated feature branches because (sometimes in near or wider future) someone else is also needing such a feature and so i easily can provide a PR and get the stuff into master....

Our problem is quite the same as linus torwalds' was in the late 90's - we need a flexible system architecture with subsystems and APIs tieing them together with the "kernel" to allow decentralized (and decentralized approved) subsystem-changes and subsystem-config without effecting other subsystems or even the "kernel". Our software base isn't really architectured; maybe it was but it has been "destroyed" by too much features and no "architecture cleanup". Maybe guy's like me have speed up this problem - so it might be the time to think about a "new guided development system for xdrip" (maybe there is one and i didn't know it).

**Comment by Navid200** [comment]
I am hoping that I can fix 193 myself.  But, I have to learn a little bit first.  That's not a problem.  But, it may take me a while.
That was just an example I was using to point out that we have open issues and I would like to close them even if I myself am not affected.
jamorham could have just compiled the code and used it for himself.  But, he chose to offer it to everyone.
xDrip serves a large portion of Diabetics on the planet.  My hat's off to jamorham and anyone else who sees value in service.

At the same time, I value talent.  Anyone who is talented and likes to contribute, I don't see why they couldn't.  If you like to add a sophisticated feature to xDrip, I don't see why not.  But, whether it is forced on everyone or not should not affect the outcome for you, should it?
Can you not have an executable that you can run after every new release to add your feature to it?

If you want to have a sports car that can also be a sports utility vehicle and also a minivan and also a chopper and also a boat, you will end up with a contraption that will not be as good as any of those alone.  xDrip is the same.  The more features we add to it, the harder it becomes to just use it as a good CGM.  We have to be careful what we do with xDrip.  I would like it to survive, and I would like the regular people to be able to use it, not just technical people.

**Comment by gruoner** [comment]
i agree with u: we got a (in germany we call it) "eierlegende wollmichsau" - it does everything and it does it because someone needed / demanded those features and there wasn't any "better" container to let your features base on.
BUT who will decide which feature is "good" or "valueable" enough to get merged and which one isn't?? Exactly thats the point where my 4 feature branches are starting - no one decided; i does. And whenever someone else is also demanding a feature like mine i will try to address it and bring it to the master branch - BUT then decision has also be taken; who decides if two guy's demanding the feature is enough? Do we need 100 users or a 1000 or even 10'000 users? Here my thoughts concerning a dedicated feature train are starting.....

BUT all those word will not bring me to an answer of my question - xdrip will not change the way u want neither developers does and "others" might also run into this issue and might demand a feature like this....?

**Comment by tolot27** [comment]
@gruoner Maybe, it's time for you to join us at gitter. Navid and I are screening and tagging all open issues. Afterwards, we will categorize them and assign a severity and priority and maybe create some milestones. Our goal is to get a good overview and restructure the code to fix bugs efficiently and in a timely manner.

**Comment by gruoner** [comment]
gitter? how do u organize your "work" to get bugs fixed? u can contact me directly at gruoner-github@gmx.net

**Comment by Navid200** [comment]
> BUT who will decide which feature is "good" or "valueable" enough to get merged and which one isn't?? 
In this case, it would be jamorham because this is his repository.

> who decides if two guy's demanding the feature is enough? Do we need 100 users or a 1000 or even 10'000 users? Here >my thoughts concerning a dedicated feature train are starting.....

A governing body (in this case, the developer headed by jamorham) makes such decisions.  It is a thankless job because no matter what you choose to do, someone is going to complain.  Sometimes, there isn't a right and wrong decision; but, you have to make a choice regardless of knowing that a portion of user community is going to whine.  

> BUT all those word will not bring me to an answer of my question - xdrip will not change the way u want neither developers does and "others" might also run into this issue and might demand a feature like this....?

I know that.
I hate it that when I drive my car, I have to push a button to lower the window.  I wish I could just rotate a lever like before for that.  But, I cannot because all car manufacturers have chosen to go that route.  So, I put up with it.  I am not going to build my own car just because I don't like how the windows operate!

Likewise, there are things I don't like in xDrip.  I communicate my wishes.  But, if I don't get my way, life goes on.  
If there is a ever a better alternative, I will switch.  I have no obligation to use xDrip.
xDrip has served me well for years and I hope that it will continue to evolve and I can continue to use it and hopefully even contribute.

**Comment by gruoner** [comment]
yeah - what else...? i want to know which way to go - u'r right (esp. the metaphor with u'r car and lever) but my issue is real and i need (and have) an optimal solution for that. The only question i have is the way to achieve the goal. 
U'r "wish" is to resign the optimization; which collides with my "wish". So it will become realized - at least in my train.
The question is: is there any other benefit someone might leverage from or can i merge it into my own branch and will never merge it to master (there are some feature i decided never to merge to master)?

**Comment by Navid200** [comment]
Everything I have posted have been my opinion.  
And thank you for the discussion.  I have learned from you.

I think jamorham is the right person to answer you.  Because I neither have the authority nor the experience to advise you in that sense.

**Comment by k2s** [comment]
First of all, @gruoner, thank you for you great contribution with multiple insulin. Your ideas and skills are great contribution to xDrip.

xDrip is nice base. The code quality is solid. xDrip is very open to extension. Functionality is overwhelming. New activities in ticket and translation management are visible and were needed. @jamorham is a rigorous, but open minded keeper of the code and I am happy to read that he doesn't want to split xDrip development into multiple streams.

There is huge variability of users and xDrip is currently best suited for "adult, technically skilled, english speaking, using pump and Dexcom". I am on different boat "parent, with relatively young child, non-english, using pen and Libre2". The multiple insulin functionality was a must for me. I would welcome easier treatment entry, easier configuration and Libre2 whole house coverage. I am so thankful xDrip for what it gives me and there is so much I would like to see improved and added. Can't imagine that xDrip should be closed from further improvements and new features. 

There can't be anything wrong to listen to user requests, bug reports, to discuss priorities and ways how to move forward (which I understand this ticket is also about). Users should not be bent out that they demand is not important and skilled contributors should be handled with respect - they are the only source of solutions. Developer community around xDrip is driven by own needs. Contribution to xDrip is not a passion, it is a must for each of us to do the best for our health or health of our most loved. Developers invest time and skills and share the results with others. If somebody really needs fix or feature and has no skills, he can hire any programmer in the world and donate the result to xDrip. 
Refusal of contribution will not redirect resource to other tickets.

**Back to ticket:**

* before I read this ticket I was trying to find a solution how to use auto config QR code to share partial config presets
* it is very difficult to give support for xDrip becasue it always ends with need of screenshots, written text and improvements of instructions
* an example is to setup the new Libre Bluetooth functionality - it requires many settings and if you miss something it will not work
* config file contains over 500 entries! they are necessary, it is impossible to say "cut them to half" or "don't add more of them"
* current "share all settings over QRcode" is using dedicated, close sourced server, where the settings are stored in encrypted form
* because of this finding I would say, @gruoner, let us  
  * extend Nightscout to support exchage of settings, archive it there and also to backup xDrip database
  * try to keep the implementation independent from NS - easy to run as standalone web server
  * keep the current server used for config exchange as is, but make it possible to enable the new NS solution for this task
  * QR code should support (maybe it already does, was looking only very quickly into the code) reading from different URL, not only the hardcoded one
* I see problem in WebDAV, that user will need to setup/configure additional service
* possibility to share publicly partial configs would make support easier

I will be more then happy to collaborate on this with you.

**Comment by gruoner** [comment]
@k2s: glad to see that i'm not alone in this boat - same on my side: young child, non-native english speaker, 3 pens (with a more complicated therapy schema) and dexcom; but with a wife with very high demands on data gouvernance and thus a local nightscout server, VPNs and a local sophisticated reporting engine together with a self-learning neural net for BG-prediction and insulin profile adoption....
I need to have ALL information influencing the BG in nightscout server to feed my neural algorithms - thus a fully functional diabetic diary was a must and i'm already working on the multicarbs feature (it is much more complicated i ever would expect)

This feature is just for convenience - last month by child crashed the mobile (display was spidered and i can't see anything but it was already running) AND to switch to the replacement device i need to know the actual config of the dexcom transmitter and some other config. i can't even do a export via sd-card or QR-code (because display was spidered); i want to have it "downloadable" from some kind of repo.
I had have seen two options - extend nightscout server or "simple" up/download via webdav.
As far as i can interpret u'r vote u suggest extending the nightscout server AND implement an even more simple webserver-upload.
Extending nightscout server isn't that much thing but both patches have to be merged and this costs time.
I don't know any more simple web upload protocol than webdav but we might use smtp/email - this might result in lots of emails but emailing is really simple. We can collect all changes to preferences within a hour or so and send it to a mailbox configured by the user. This isn't a downloadable config but will also fit my biggest issue and i can use the import function to get it online on the replacement device.....

**Comment by Navid200** [comment]
I'm looking at this from a different angle now.
There is a post on facebook describing the child's phone having broken beyond repair.
They now have a new phone and want to know the easiest way to set it up.
If uploading the settings, to Nightscout, was possible, and if they had Nightscout, they could transfer the settings to the new phone and everything, the alerts, the glucose levels, the sharing settings, etc. would be all set.

My problem is I still don't understand how storage space on Nightscout is used.
I have been told that no matter how often you upload something to Nightscout, it is going to be replicated once every 5 minutes.  So, I don't understand how that can be sustainable if we keep adding more and more data to Nightscout.
If that is not an issue, I would like this for the purpose I have described.  Now, if someone else likes to use the same capability for constantly changing the settings, I don't care.
But, will this occupy space on Nightscout that otherwise could have been used to store BG readings for a longer period of time?

I have several questions:

Exactly how storage space on Nightscout is allocated to different things?

What will be the impact on bandwidth and data usage considering it's not just a single pair of blood glucose and time we are uploading; it's many many settings.

If the phone is broken and we are setting up a new phone, what's the easiest way to transfer the settings to the new phone?

Is there an easier way to accomplish what I need rather than this new feature for xDrip? 

<br/>
<br/>

@sulkaharo would you be kind again to help me understand how Nightscout allocates space?
Does such capability already exist in Nightscout for any other app?  I mean can a user just establish connectivity to Nightscout from their app and easily transfer all the app settings, already saved on Nightscout, to their app?

This is what my Nightscout looks today:
![Screenshot 2021-08-09 105618](https://user-images.githubusercontent.com/51497406/128729254-6ff8a9f0-eee3-40af-b982-b71fa058d799.png)

If I understand it correctly, after 3 years, I have occupied 25% of the total storage I have in my free account.
So, my simple calculation tells me I can go 9 more years until that reaches 100%.  Is that right?
I am only storing glucose readings from xDrip and nothing else.
If we add xDrip settings to this, will that fill the same space in 1 year instead of 9, or 8 instead of 9, or 8 years and 11 months instead of 9 years?  I have no perspective?  Is there a way to know that?

When I get close to that, I suppose I need to delete some data to free up space.
Can I delete data by data type or only by date?
In other words, will I be able to delete all the xDrip settings saved and leave all the blood glucose readings alone?

**Comment by gruoner** [comment]
@Navid200 lots of questions and i don't have valuable answers for u:
**Storage**: as we just store all current config info there isn't that much consumption. Keep in mind: BG and treatment data are already stored. I assume, it will take just a few hundred kilobytes
**bandwith**: that's a really more complicated thing as these config information must either be transferred  incrementally or monolithic as one block; maybe (depends on the nightscout API) we can split it up in more than just one block
**restore-procedure**: "someone" has to configure the access parameters to the nightscout server - whenever u have done configuring nightscout-upload, u know what i mean. We have to find a way to restore all from scratch just by presenting a blank config file and some kind of 2D barcode (QR-code) holding the nightscout-link....

**Comment by gruoner** [comment]
i'm afraid that's the wrong answer for the correct question - one has to find the cause why complete databases are repeatedly droped.... for sure, this feature will lower the pain when a db drop occurs but that can't be the solution.....

i'm currently working on a webdav backup/restore feature (it's useless to upload all of xdrip's db entities (tables) to dedicated nightscout mongo-collections, so i will use the existing backup/restore functions and add just a webdav protocol handling).
But this might take some weeks to accomplish (its vacation time).....

**Comment by Navid200** [comment]
I don't think I ever said that was the solution.

**Comment by gruoner** [comment]
opened a PR #2269

**Comment by Navid200** [comment]
xDrip can now create a backup on Google Drive.  And it can restore from it.
Does that address the need?
https://navid200.github.io/xDrip/docs/GoogleDriveBackup.html

**Comment by gruoner** [comment]
no it doesn't because i use an android mobile without any google apps
BUT this isn't any problem because i implemented a webdav upload/restore feature
